[SYCL][Graph] Add exceptions on using spec constants, kernel bundles and reductions

Bensuo opened the pull request:
- Add exceptions when using spec constants, kernel bundles and reductions
- Unit tests for spec constants and kernel bundles.
- Refactored handler throwing code to remove templates.
- Moved and renamed unsupported features enum to graph header.

EwanC commented on sycl/include/sycl/handler.hpp:
```diff
@@ -3382,18 +3383,20 @@ class __SYCL_EXPORT handler {
 
-  template <ext::oneapi::experimental::detail::SyclExtensions ExtensionT>
-  void throwIfGraphAssociated();
+  void throwIfGraphAssociated(
+      ext::oneapi::experimental::detail::UnsupportedGraphFeatures Feature)
+      const;
```
Reckon this will need an update to the `sycl/test/abi/sycl_symbols_linux.dump` test. I can push a windows test update.

Bensuo replied:
Updated now, thanks.

EwanC commented on sycl/source/handler.cpp:
```diff
@@ -1109,6 +1110,9 @@ void handler::ext_oneapi_signal_external_semaphore(
 
+  throwIfGraphAssociated(ext::oneapi::experimental::detail::
+                             UnsupportedGraphFeatures::sycl_kernel_bundle);
+
```
Well remembered about kernel bundles, when we update the list of unsupported features for the release we should mention kernel bundles.

EwanC reviewed (approved):
LGTM, I'm hopeful that removing templates from `handler::throwIfGraphAssociated` will solve the Windows CI fail, but I can try verify that locally.

EwanC commented:
> I'm hopeful that removing templates from `handler::throwIfGraphAssociated` will solve the Windows CI fail, but I can try verify that locally.

Yup, this branch does fix the linking error on Windows

Bensuo commented:
@EwanC I've also pushed changes adding exceptions for Reductions here to make use of the refactored error handling code. Tagging for re-review, sorry!

EwanC commented on sycl/unittests/Extensions/CommandGraph.cpp:
```diff
@@ -1139,6 +1154,54 @@ TEST_F(CommandGraphTest, Memcpy2DExceptionCheck) {
+      },
+      sycl::exception);
+}
+
```
Could have a reduction test here verifying the exception thrown, E2E reduction tests don't check how the test fails.

Bensuo replied:
Done!

EwanC reviewed (approved):
Reductions LGTM, minor comment at unittesting it

mfrancepillois reviewed (commented):
In a previous PR, I changed the parametrized throwIfGraphAssociated function into a template function to be in line with other throw functions used by handle (as suggest by @reble in his comment https://github.com/reble/llvm/pull/276#discussion_r1281353861 ). 
I understand that changing back the template function into a parametrized function fixes the Windows compilation issue, but we should be aware that less consistent with the other throw functions of this file (and it also may impact performances as it makes the inlining of this function less intuitive for the compiler?)

reble reviewed (approved):
LGTM

Bensuo commented:
> In a previous PR, I changed the parametrized throwIfGraphAssociated function into a template function to be in line with other throw functions used by handle (as suggest by @reble in his comment [#276 (comment)](https://github.com/reble/llvm/pull/276#discussion_r1281353861) ). I understand that changing back the template function into a parametrized function fixes the Windows compilation issue, but we should be aware that less consistent with the other throw functions of this file (and it also may impact performances as it makes the inlining of this function less intuitive for the compiler?)

I've changed this back to a template but kept the definition in the header to avoid the template specializations 👍

mfrancepillois reviewed (approved):
LGTM
